Shopify/cli
# New Result type for handling errors

### WHY are these changes introduced?
Fixes https://github.com/Shopify/shopify-cli-planning/issues/312
Follow-up: https://github.com/Shopify/cli/pull/319

### WHAT is this pull request doing?
- New `Result` type added. Used as a return type to indicate the operation has suceded or should return an error
- Three utility methods for creating an ok and error response, and to get the value of an ok response or throw a fatal exception (Bug, Abort, AbortSilent) otherwise
- Used new type in `automaticMatchmaking`

TBD: It also should be added to `manualMatchIds` in case the new type is approved

### How to test your changes?
- Run `yarn shopify app deploy` with a remote app that matches local extensions and then with another that does not match to see the abort error message
